Model classification Logistic Regression


In [53]:
# import library
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [54]:
# Importing Dataset
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
dataset = pd.read_csv('diabetes_prediction_dataset.csv')
dataset['gender'] = lab.fit_transform(dataset['gender'])
X = dataset.iloc[:, :-1]
y = dataset.iloc[:, -1]

In [55]:
dataset

gender   age  hypertension  heart_disease smoking_history    bmi  \
0           0  80.0             0              1           never  25.19   
1           0  54.0             0              0         No Info  27.32   
2           1  28.0             0              0           never  27.32   
3           0  36.0             0              0         current  23.45   
4           1  76.0             1              1         current  20.14   
...       ...   ...           ...            ...             ...    ...   
99995       0  80.0             0              0         No Info  27.32   
99996       0   2.0             0              0         No Info  17.37   
99997       1  66.0             0              0          former  27.83   
99998       0  24.0             0              0           never  35.42   
99999       0  57.0             0              0         current  22.43   

       HbA1c_level  blood_glucose_level  diabetes  
0              6.6                  140         0  
1              6.6                   80         0  
2              5.7                  158         0  
3              5.0                  155         0  
4              4.8                  155         0  
...            ...                  ...       ...  
99995          6.2                   90         0  
99996          6.5                  100         0  
99997          5.7                  155         0  
99998          4.0                  100         0  
99999          6.6                   90         0  

[100000 rows x 9 columns]

In [56]:
# Periksa data kosong
X.isnull().any()

gender                 False
age                    False
hypertension           False
heart_disease          False
smoking_history        False
bmi                    False
HbA1c_level            False
blood_glucose_level    False
dtype: bool

In [57]:
y.isnull().any()

False

In [58]:
# Membuang data yang kosong
X = X.dropna()
y = y.dropna()

In [59]:
# membuat index menjadi berurutan setelah membuang data yang kosong
X.reset_index(drop=True, inplace=True)
y.reset_index(drop=True, inplace=True)

In [60]:
# Periksa data duplikat
X.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
99995     True
99996    False
99997    False
99998    False
99999    False
Length: 100000, dtype: bool

In [61]:
X.drop_duplicates()

gender   age  hypertension  heart_disease smoking_history    bmi  \
0           0  80.0             0              1           never  25.19   
1           0  54.0             0              0         No Info  27.32   
2           1  28.0             0              0           never  27.32   
3           0  36.0             0              0         current  23.45   
4           1  76.0             1              1         current  20.14   
...       ...   ...           ...            ...             ...    ...   
99994       0  36.0             0              0         No Info  24.60   
99996       0   2.0             0              0         No Info  17.37   
99997       1  66.0             0              0          former  27.83   
99998       0  24.0             0              0           never  35.42   
99999       0  57.0             0              0         current  22.43   

       HbA1c_level  blood_glucose_level  
0              6.6                  140  
1              6.6                   80  
2              5.7                  158  
3              5.0                  155  
4              4.8                  155  
...            ...                  ...  
99994          4.8                  145  
99996          6.5                  100  
99997          5.7                  155  
99998          4.0                  100  
99999          6.6                   90  

[96055 rows x 8 columns]

In [62]:
# Periksa data kolom 'age'
column_name = 'age'
non_integer_data = X[column_name].apply(lambda x: not isinstance(x, int))

In [63]:
X

gender   age  hypertension  heart_disease smoking_history    bmi  \
0           0  80.0             0              1           never  25.19   
1           0  54.0             0              0         No Info  27.32   
2           1  28.0             0              0           never  27.32   
3           0  36.0             0              0         current  23.45   
4           1  76.0             1              1         current  20.14   
...       ...   ...           ...            ...             ...    ...   
99995       0  80.0             0              0         No Info  27.32   
99996       0   2.0             0              0         No Info  17.37   
99997       1  66.0             0              0          former  27.83   
99998       0  24.0             0              0           never  35.42   
99999       0  57.0             0              0         current  22.43   

       HbA1c_level  blood_glucose_level  
0              6.6                  140  
1              6.6                   80  
2              5.7                  158  
3              5.0                  155  
4              4.8                  155  
...            ...                  ...  
99995          6.2                   90  
99996          6.5                  100  
99997          5.7                  155  
99998          4.0                  100  
99999          6.6                   90  

[100000 rows x 8 columns]

In [64]:
# Definisikan map dictionary untuk 'smoking_history'
mapping = {
    "No Info": 0,
    "never" : 1,
    "former" : 2,
    "not current" : 3,
    "ever" : 4,
    "current" : 5,
}
X['smoking_history'] = X['smoking_history'].replace(mapping)
dataset.head(5)


gender   age  hypertension  heart_disease smoking_history    bmi  \
0       0  80.0             0              1           never  25.19   
1       0  54.0             0              0         No Info  27.32   
2       1  28.0             0              0           never  27.32   
3       0  36.0             0              0         current  23.45   
4       1  76.0             1              1         current  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0

In [65]:
y = np.array(y).reshape(len(y),1)

In [66]:
# Bagi dataset menjadi data latih dan data uji
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test =train_test_split(X,y,test_size=0.25,random_state=42)


In [67]:
X_test

gender   age  hypertension  heart_disease  smoking_history    bmi  \
75721       0  13.0             0              0                0  20.82   
80184       0   3.0             0              0                0  21.00   
19864       1  63.0             0              0                2  25.32   
76699       0   2.0             0              0                1  17.43   
92991       0  33.0             0              0                3  40.08   
...       ...   ...           ...            ...              ...    ...   
21271       0  25.0             0              0                0  27.32   
34014       0  42.0             0              0                0  20.05   
81355       0  56.0             0              0                0  27.32   
65720       1  40.0             0              0                1  29.06   
11627       0  24.0             0              0                1  22.99   

       HbA1c_level  blood_glucose_level  
75721          5.8                  126  
80184          5.0                  145  
19864          3.5                  200  
76699          6.1                  126  
92991          6.2                  200  
...            ...                  ...  
21271          5.8                  145  
34014          4.0                  200  
81355          6.5                   90  
65720          5.0                  200  
11627          6.2                  159  

[25000 rows x 8 columns]

In [68]:
print(y_train)

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


In [69]:
print(y_test)

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


Feature Scaling

In [70]:
# regresi model yg membutuhkan fiture scalling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train= sc.fit_transform(X_train)
X_test= sc.transform(X_test)

In [71]:
print(X_train)

[[-8.42792712e-01 -6.65004023e-01 -2.85922318e-01 ...  2.96341795e-04
   1.00248870e+00  4.90070554e-01]
 [-8.42792712e-01 -1.06493670e+00 -2.85922318e-01 ... -8.25681898e-01
  -1.42421796e+00  1.52093063e+00]
 [ 1.18445082e+00 -3.98382239e-01 -2.85922318e-01 ...  2.96341795e-04
  -4.90869245e-01 -1.42438388e+00]
 ...
 [ 1.18445082e+00  1.55043529e-03 -2.85922318e-01 ... -1.78539314e-01
   2.55809727e-01 -1.30166244e+00]
 [-8.42792712e-01 -2.20634384e-01 -2.85922318e-01 ... -3.57374969e-01
   6.29149212e-01  4.90070554e-01]
 [-8.42792712e-01 -8.42751878e-01 -2.85922318e-01 ...  1.01838621e-01
  -4.90869245e-01  5.14614842e-01]]


In [72]:
print(X_test)

[[-8.42792712e-01 -1.28712152e+00 -2.85922318e-01 ... -9.84815320e-01
   2.55809727e-01 -2.95346649e-01]
 [-8.42792712e-01 -1.73149116e+00 -2.85922318e-01 ... -9.57535305e-01
  -4.90869245e-01  1.70994816e-01]
 [ 1.18445082e+00  9.34726676e-01 -2.85922318e-01 ... -3.02814939e-01
  -1.89089232e+00  1.52093063e+00]
 ...
 [-8.42792712e-01  6.23667929e-01 -2.85922318e-01 ...  2.96341795e-04
   9.09153827e-01 -1.17894100e+00]
 [ 1.18445082e+00 -8.73234924e-02 -2.85922318e-01 ...  2.64003156e-01
  -4.90869245e-01  1.52093063e+00]
 [-8.42792712e-01 -7.98314914e-01 -2.85922318e-01 ... -6.55939580e-01
   6.29149212e-01  5.14614842e-01]]


In [73]:
# Training the Logistic Regression model on the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(random_state=0)

In [74]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


Making the Confusion Matrix

In [75]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test,y_pred ) #y_test bida dignti dgn y_true
print(cm)

[[22649   201]
 [  832  1318]]


In [76]:
accuracy_score(y_test, y_pred)

0.95868

2. Training the KNN model on the Training set

In [77]:
from sklearn.neighbors import KNeighborsClassifier
classifier2 = KNeighborsClassifier(n_neighbors=5, p=2, metric='minkowski')
classifier2.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


KNeighborsClassifier()

In [78]:
y_pred2 = classifier2.predict(X_test)
print(np.concatenate((y_pred2.reshape(len(y_pred2),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


Making the Confusion Matrix

In [79]:
cm2 = confusion_matrix(y_test,y_pred2)
print(cm2)

[[22701   149]
 [  848  1302]]


In [80]:
accuracy_score(y_test,y_pred2)

0.96012

3.Training the SVM model on the Training set

In [81]:
from sklearn.svm import SVC
classifier3 = SVC(kernel='linear', random_state = 0)
classifier3.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(kernel='linear', random_state=0)

In [82]:
y_pred3 = classifier3.predict(X_test)
print(np.concatenate((y_pred3.reshape(len(y_pred3),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


In [83]:
cm3 = confusion_matrix(y_test,y_pred2)
print(cm3)

[[22701   149]
 [  848  1302]]


In [84]:
accuracy_score(y_test,y_pred3)


0.9586

Kesimpulan
1. classification Logistic Regression = 0.95868
2. classification K-NN = 0.96012
3. classification SVM dengan linear kernel = 0.9586
